### PR TITLE
Adds media url provider

### LINKED ISSUE #4942: Create media Url provider
Relates to #4941 
Relates to #4940

Just like we have for content, we need to have a media URL provider which can be replaces (just like content url providers). This is an important step towards being able to properly support CDNs and things like Blob media providers. 

If media is stored in blob storage and is pushed to a CDN, we need to be able to return the CDN media URL from a media URL provider.

This will be a bit more tricky than the content url provider because media can exist on both content and on media and can exist multiple times per content or media item.

### PR DESCRIPTION
**Pull request by rasmusjp**
fixes #4942 

### Breaking changes

* `UrlProvider` constructors now takes an `IEnumerable<IMediaUrlProvider>`
* `UmbracoContextFactory` constructor now takes a `MediaUrlProviderCollection`

### Things to test

* `media.Url` returns the url for the `umbracoFile` property,
* `media.UrlAbsolute()` returns the absolute url for the `umbracoFile` property.
* `media.GetCropUrl(200, 200)` returns the crop url

Add a `Umbraco.UploadField` and a `Umbraco.ImageCropper` property to a content item and test the following both as invariant and variant

* `content.MediaUrl('propertyAlias')` returns the url for the media in the `propertyAlias` property
* `content.MediaUrlAbsolute('propertyAlias')` returns the absolute url for the media in the `propertyAlias` property
* `content.MediaUrl('propertyAlias', "culture")` returns the url for the media in the `propertyAlias` property in the specified culture
* `content.MediaUrlAbsolute('propertyAlias', "culture")` returns the absolute url for the media in the `propertyAlias` property in the specified culture
* `content.GetCropUrl(200, 200 propertyAlias: "propertyAlias")` returns the crop url

Crete a new implementation `IMediaUrlProvider` which always returns a full url and add it as the first one in the `MediaUrlProviderCollection` e.g.

```csharp
public class CdnMediaUrlProvider : DefaultMediaUrlProvider
{
    public override UrlInfo GetMediaUrl(UmbracoContext umbracoContext, IPublishedContent content, string propertyAlias,
        UrlProviderMode mode, string culture, Uri current)
    {
        var mediaUrl = base.GetMediaUrl(umbracoContext, content, propertyAlias, UrlProviderMode.Relative, culture, current);
        if (mediaUrl == null)
            return null;

        if (mediaUrl.IsUrl == false)
            return mediaUrl;

        var cdnBaseUrl = "https://my.cdn";

        return UrlInfo.Url(cdnBaseUrl + mediaUrl.Text, culture);
    }
}

public class CdnComposer : IUserComposer
{
    public void Compose(Composition composition)
    {
        composition
            .MediaUrlProviders()
            .Insert<CdnMediaUrlProvider>();
    }
}
```

and test that all the above url's now is prefixed with `https://my.cdn`

### CODE REVIEW

**Comment by Shazwazza** [comment]
nice, haven't had time to fully look at this one but just wanted to ping @zpqrtbnk since he'll also have an interest in this one too.

**Comment by sitereactor** [comment]
What does this implementation mean for regular every day use of Umbraco? 
1. Anything that needs to be done differently to get the media path/url through this provider?
2. Any scenarios that are not covered? Like content.GetPropertyValue("propertyAlias")

Just trying to get an idea of what would work and potentially what wouldn't when you have a CDN provider in place. Is there potential for getting media Urls without the CDN prefix.

**Comment by Shazwazza** [comment]
@sitereactor not sure if you've read all of the related issues? This is what this is fixing: https://github.com/umbraco/Umbraco-CMS/issues/4942, this is the epic that contains it: https://github.com/umbraco/Umbraco.Private/issues/283, this is the spike with some research done https://github.com/umbraco/Umbraco-CMS/issues/4943
... otherwise please keep asking questions :)

**Comment by sitereactor** [comment]
Okay, I will try to think of another way to ask my question so you will understand. From your reply I gather its not clear 😅

**Comment by sitereactor** [comment]
How about this:

Given all the tasks/subtasks mentioned in the #283 Epic are fully implemented. Will the use of getting a path/url to a media item be completely seemless to the user or is there anything that would require the developers to use different methods?

Say I'm on version 8.0.x and all the tasks from the Epic is implemented in 8.x and I upgrade to this version and configure a media provider to add a CDN to the site. Will everything Just Work™️ ?

**Comment by sitereactor** [comment]
Great, thank you.

**Comment by rasmusjp** [comment]
For media `.Url` and `.GetCropUrl()` will use the url provider instead of just returning the `umbracoFile` property value, but if you have a upload/image cropper field on your content and uses `content.Value()` it will return the property value (either a `string` or a `ImageCropperValue`) and it's not using the url providers, so you'll need to use `content.MediaUrl("")` instead.

I guess we could make `content.Value()` work by changing the value converters for upload and image cropper, but not sure what would be the best approach here, maybe @zpqrtbnk has some input?

**Comment by sitereactor** [comment]
What about from a package developers perspective? Anything special they'd need to be aware of? Something around how to handle media paths within content.
Sorry about my lack of insight here. Just trying to understand.

File: src/Umbraco.Web/Routing/DefaultMediaUrlProvider.cs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+                    break;
+                case UrlProviderMode.Relative:
+                case UrlProviderMode.Auto:
+                    uri = new Uri(path, UriKind.Relative);
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(mode));
+            }
+
+            return uri.Rewrite(UriUtility.ToAbsolute(uri.GetSafeAbsolutePath()));
```
Comment (zpqrtbnk) [inline src/Umbraco.Web/Routing/DefaultMediaUrlProvider.cs]: Not sure this does what we want. Calling `ToAbsolute` on an already absolute url is... going to do weird things? Looking into it.

**Comment by zpqrtbnk** [comment]
@rasmusjp I have pushed a commit that does some minor fixes, want to check I haven't done anything stupid?

Now running tests.

As for the behavior of `Value()` etc - see #5170 - we're going to refactor `Url()` and more generally the `IPublishedContent` interface in 8.1, so I *think* some of the methods defined in this PR will change again with #5170.

Ideally, `Value()` returns "the value" whatever that is, and `Url()` returns "the url". So I'd rather not use `Value()` to return a Url. And by default `Url()` should work with... `umbracoFile` if present, so ppl don't have to supply the property alias, by default. But can, if they want to override.

So in a view... `` should "just" work - and if a dev replaces the media url provider, she can control what `.Url()` returns here.

Making sense?

**Comment by ronaldbarendse** [comment]
How do you get an absolute crop URL? `GetCropUrl()` should probably get an additional parameter to specify the `UrlMode` that is passed to `MediaUrl` for this, right?

https://github.com/umbraco/Umbraco-CMS/blob/3bfd9b71e290354744d677440983ecd06c5c0788/src/Umbraco.Web/ImageCropperTemplateExtensions.cs#L133

**Comment by ronaldbarendse** [comment]
And all `GetCropUrl()` extension methods should probably also allow specifying a culture, so all `MediaUrl()` overloads are covered:

https://github.com/umbraco/Umbraco-CMS/blob/3bfd9b71e290354744d677440983ecd06c5c0788/src/Umbraco.Web/PublishedElementExtensions.cs#L207

**Comment by ronaldbarendse** [comment]
BTW: adding support for culture variant media types will be required first: https://github.com/umbraco/Umbraco-CMS/issues/4771.

**Pull request merged**